[js] Adding logger documentation (#1321)
SeleniumHQ/seleniumhq.github.io: website_and_docs/content/documentation/webdriver/troubleshooting/logging.en.md

Each language adopts a distinctly different approach to logging information about the activity
of the program.

### 1. Getting a logger

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}

>   <p>Ruby uses a custom implementation of the default `Logger` class with some interesting additional features.</p>

Reviewer: <p>Ruby uses a custom implementation of the default `Logger` class with some interesting additional features.</p>
```
logger = Selenium::WebDriver.logger
```

  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  <pre>
  const logging = require('selenium-webdriver/lib/logging')
  logger = logging.getLogger('webdriver')</pre>
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

### 2. Logger Level
Logger level helps to filter out logs based on their severity.

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}

>   <p>Levels are: <i>`:debug`, `:info`, `:warn`, `:error`, `:fatal`.</i> Default is <i>`:warn`</i>.</p>

Reviewer: These are code symbols, I wouldn't put them in italics

  <p>To change the level of the logger:</p>
  <pre>

>   Selenium::WebDriver.logger.level = :fatal</pre>

Reviewer: ```suggestion
  Selenium::WebDriver.logger.level = :info</pre>
```

  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  <p>Levels are: <i>`OFF`, `SEVERE`, `WARNING`, `INFO`, `DEBUG`, `FINE`, `FINER`, `FINEST`, `ALL`</i>. Default is <i>`OFF`</i>.</p>
  <p>To change the level of the logger:</p>
  <pre>
  logger.setLevel(logging.Level.INFO)</pre>
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

> **WARN**

Reviewer: Instead of the Ruby constants, let's make these headers normal English:
**Actionable Items**

> Warnings include everything we want users to be aware of by default. This is mostly used

Reviewer: ```suggestion
Things are logged as warnings if they are something the user needs to take action on. This is mostly used
```

  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}
  <p>Ruby logs deprecations as warnings, specifying what is changing, what needs to be
  used instead. It may include additional messages, and always includes an ID.</p>

  <p>For example:</p>
  <blockquote>2022-12-24 16:07:09 WARN Selenium [DEPRECATION] [:jwp_caps] \`Capabilities#version=\` is deprecated. Use \`Capabilities#browser_version=\` instead.</blockquote>

  <p>Because these items can get annoying, we've provided an easy way to turn them off.</p>

  <p>To turn off a specific warning, set the ID to ignore:</p>
  <pre>
  Selenium::WebDriver.logger.ignore(:jwp_caps)</pre>
  <br>
  <p>It accepts an `Array` to turn off multiple IDs:</p>
  <pre>
  Selenium::WebDriver.logger.ignore(%i[jwp_caps pause pauses])</pre>
  <br>
  <p>To turn off all *deprecation* notices:</p>
  <pre>
  Selenium::WebDriver.logger.ignore(:deprecations)</pre>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

> **INFO**

Reviewer: ```suggestion
**Useful Information**
```

<pre>
  Selenium::WebDriver.logger.ignore(%i[jwp_caps pause pauses])
</pre>
<br>
<p>To turn off all <i>deprecation</i> notices:</p>
<pre>
  Selenium::WebDriver.logger.ignore(:deprecations)
</pre>
<br>
<p><b>INFO</b></p>

<p>This is where the most useful information gets logged. Selenium logs the endpoints and payloads
sent to and received from the driver or server. This is a great way to see what Selenium is actually
doing under the hood, and can be used to determine if it is Selenium code or driver code that
is causing a problem. (Unfortunately, we can't blame the driver if Selenium is sending incorrect syntax).
</p>

<p><b>DEBUG</b></p>

<p>This is less useful information where we log things about the servers and the sockets, and header information, etc.
Debug mode is set if either <i>`$DEBUG`</i> is true or <i>`ENV['DEBUG']`</i> has a value.</p>
{{< /tab >}}

{{< tab header="JavaScript" >}}
<p>JS has implemented logging APIs. It has also added LogManager to maintain a collection of loggers.</p>
<p><b>1. Getting a logger can be done in two ways:</b></p>
<pre>
  const logging = require('selenium-webdriver/lib/logging')
</pre>
<br>

> <p>First method using manager</p>

Reviewer: Is the only reason to use the manager if you need more than one logger?

Author: I looked further into LogManager. Though it helps us to store loggers it is not required in our case. We don't need to explicitly use a LogManager. Selenium maintains a LogManager when it starts. I have cleaned the approach in the new commit.

<pre>
  mgr = new logging.LogManager()
  logger = mgr.getLogger('')
</pre>
<br>
<p>Second method without manager</p>
<pre>
  logger = logging.getLogger('')
</pre>
<br>
<p>Loggers class uses hierarchical, dot-delimited namespaces (e.g. "" > "webdriver" > "webdriver.logging"). For example, the following code will give a hierarchy of loggers:</p>
<pre>

>   logger = mgr.getLogger('foo.bar.baz')

Reviewer: Can we use a real Selenium example instead of `foo.baz.bar` I'm still not sure I get what this is doing.

Author: No need to use this as we are not creating a logger or using a manager. We are just getting the already created logger 'webdriver'.

</pre>
<br>
<p>"" (parent of) > "foo" (parent of) > "foo.bar" (parent of) > "foo.bar.baz" (=logger)</p>

<p><b>2. Logger Level:</b></p>

<p>The basic levels for JS loggers are: <i>`OFF`, `SEVERE`, `WARNING`, `INFO`, `DEBUG`, `FINE`, `FINER`, `FINEST`, `ALL`</i>.<br>

> To log messages at the given level pass the level in the <i>`log()`</i> method:</p>

Reviewer: How to log messages is important for writing Selenium code, but this page is intended to be for users who need to look at the logs coming from Selenium.

Author: Thank you for the clarity. This helped much. I was focussing on users' logging too. But now I have changed the approach to only target logs which Selenium generates.

<pre>
  logger.log(logging.Level.INFO, 'This is an info message')
</pre>
<br>
<p>You can also use the instance methods to target logs of a particular level:<p>
<pre>
  logger.finest('this is the finest message')
  logger.finer('this is a finer message')
  logger.info('this is an info message')
  logger.warning('this is a warning message')
  logger.severe('this is a severe message')
</pre>
<br>
<p>By default the logger's level is <i>`Level.OFF`.</i> To change the level of the logger:</p>
<pre>
  logger.setLevel(logging.Level.INFO)
</pre>
<br>
<p><b>3. Log Output:</b></p>

> <p>As opposed to ruby which by default sends logs to console in stdout, JS has the option to add a handler. The handler will be invoked for each message logged with this instance, or any of its descendants.</p>

Reviewer: So what is the default output for JS? Do you have to add a handler?

Author: There is no default output for JS. You have to use a method `installConsoleHandler()` to send the logs to the console.
Storing in a file is also not possible as of now.

Reviewer: Wait, there is no file handler that can be used?

Author: At present no API is there to store logs in a file like ruby does. But as discussed in TLC, I will implement the API for the same later.

<pre>
  logger.addHandler(callback)
</pre>
<br>

> <p>JS has provided a method to add console handler to the given logger. The console handler will log all messages using the JavaScript Console API:</p>

Reviewer: It is probably an issue of me not understanding JavaScript, but what is the difference between a handler and a console handler?

Author: A handler will take a callback function and it will be invoked when the event(log) occurs. Console handler is a special type of handler which will display the logs on console in a fixed format. 
https://github.com/SeleniumHQ/selenium/blob/trunk/javascript/node/selenium-webdriver/lib/logging.js#L526-L559

But for user's perspective to display selenium logs, we will use `installConsoleHandler()`.

<pre>
  logging.addConsoleHandler(logger)
</pre>
<br>
<p>Similarly, to add console handler to the root logger:</p>
<pre>
  logging.installConsoleHandler()
</pre>
<br>

> <p><b>4. Logging Preference (remote logging API):</b></p>

Reviewer: This is actually not maintained any longer, and shouldn't be included. :(

Author: Got it. Removed it in the latest commit.

Logger level helps to filter out logs based on their severity.

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}
  <p>Levels are: <code>:debug</code>, <code>:info</code>, <code>:warn</code>, <code>:error</code>, <code>:fatal</code>. Default is <code>:warn</code>.</p>
  <p>To change the level of the logger:</p>
  <pre>
  Selenium::WebDriver.logger.level = :info</pre>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  <p>Levels are: <code>OFF</code>, <code>SEVERE</code>, <code>WARNING</code>, <code>INFO</code>, <code>DEBUG</code>, <code>FINE</code>, <code>FINER</code>, <code>FINEST</code>, <code>ALL</code>. Default is <code>OFF</code>.</p>
  <p>To change the level of the logger:</p>
  <pre>
  logger.setLevel(logging.Level.INFO)</pre>
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

**Actionable Items**

> Things are logged as warnings if they are something the user needs to take action on. This is mostly used

Reviewer: ```suggestion
Things are logged as warnings if they are something the user needs to take action on. This is often used
```

  <pre>
  Selenium::WebDriver.logger.ignore(:jwp_caps)</pre>
  <br>
  <p>It accepts an `Array` to turn off multiple IDs:</p>
  <pre>
  Selenium::WebDriver.logger.ignore(%i[jwp_caps pause pauses])</pre>
  <br>
  <p>To turn off all *deprecation* notices:</p>
  <pre>
  Selenium::WebDriver.logger.ignore(:deprecations)</pre>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

**Useful Information**

This is the default level where Selenium logs things that users should be aware of but do not need to take actions on.
It presents information such as requests and responses between driver and server, payload, etc.

Different languages have different level to log information.

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}

>   Logs useful information at level: <code>:info</code>

Reviewer: ```suggestion
Because the Ruby `Logger` class only has one "debug" level, Selenium is currently using the `:info` level as a general debug mode, and `:debug` as a lower level debug mode.
To bring things in line with other languages, we are considering [changing this behavior]({{< ref "https://github.com/SeleniumHQ/selenium/issues/11797" >}}).
For now, both info and warnings are handled at the default `:warn` level.
```

  {{< /tab >}}
  {{< tab header="JavaScript" >}}

>   Logs useful information at level: <code>FINER</code>

Reviewer: ```suggestion
  Logs useful information at level: <code>INFO</code>
```

  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}
  Logs debugging details at level: <code>:info</code> and <code>:debug</code>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  Logs debugging details at level: <code>FINER</code> and <code>FINEST</code>
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

### 3. Log Output:
Logs can be displayed on `stdout` or stored in a file.

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}
  <p>By default, logs are sent to the console in `stdout`.</p>
  <p>To store the logs in a file:</p>
  <pre>
  Selenium::WebDriver.logger.output = '/path/to/selenium.log'</pre>
  <br>
  <p>Sample Output:</p>

>   <blockquote>

Reviewer: This formatting is hard to read, I think we need to treat this as a codeblock instead of a blockquote so that it uses a mono format.

  2023-02-08 12:31:23 INFO Selenium -> POST session/8b83ff54712d0a247937d045f8c8e171/url<br>
  2023-02-08 12:31:23 INFO Selenium    >>> http://127.0.0.1:9515/session/8b83ff54712d0a247937d045f8c8e171/url | {"url":"https://www.selenium.dev/selenium/web/web-form.html"}<br>
  2023-02-08 12:31:23 INFO Selenium <- {"value":null}</blockquote>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  <p>Send logs to console output:</p>
  <pre>
  logging.installConsoleHandler()</pre>
  <br>
  <p>Sample Output:</p>

>   <blockquote>

Reviewer: same as above, codeblock not blockquote

> doing under the hood, and can be used to determine if it is Selenium code or driver code that

Reviewer: This is only true for Ruby. We should move all of this to debugging info section for Ruby.
"Useful Information" section should say something like: "This is the default level where Selenium logs things that users should be aware of but do not need to take actions on."

is causing a problem. (Unfortunately, we can't blame the driver if Selenium is sending incorrect syntax).

Different languages have different level to log requests and response.

{{< tabpane langEqualsHeader=true text=true >}}
  {{< tab header="Java" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Python" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="CSharp" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
  {{< tab header="Ruby" >}}
  Logs request and response at level: <i>`:info`</i>
  {{< /tab >}}
  {{< tab header="JavaScript" >}}
  Logs request and response at level: <i>`FINER`</i>
  {{< /tab >}}
  {{< tab header="Kotlin" >}}
  {{< alert-content >}}{{< /alert-content >}}
  {{< /tab >}}
{{< /tabpane >}}

> **DEBUG**

Reviewer: ```suggestion
**Debugging Details**
```

> This is less useful information where we log things about the servers and the sockets, and header information, etc.

Reviewer: Again these words only apply for Ruby, so can be moved to the Ruby only section. Don't worry about Ruby tab too much, put the right words that will apply to all logging and I'll add some things to the Ruby tab to clarify.

Author: Though I have written a general statement wrt debug level, the question which arises is are we targeting the same `'debug'` level in all languages? Cause almost every lang has a debug level but the logs which they show may not be the same.

For example, you have already compared ruby and python [here](https://gist.github.com/titusfortner/ef2b4c4d3ebc32dd91e44d55c2d5d2d5). (https://github.com/SeleniumHQ/selenium/issues/10944#issuecomment-1423297676)

I have used the same [script](https://www.selenium.dev/documentation/webdriver/getting_started/first_script/#putting-everything-together) (JS version) and found that no logs are showing for DEBUG level.

What do you suggest in this case? @titusfortner

Reviewer: Yeah, the suggestion is to have the headline be "Debugging details" or similar and then describe which settings apply for that language. In Java & JS that's going to be `FINER` for most things, right? For Ruby, right now, it's both `:info` and `:debug` (and I created an issue to make them match here https://github.com/SeleniumHQ/selenium/issues/11797)
